dimod/generators/magic_square.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2023 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from itertools import product
+
+from dimod.constrained import ConstrainedQuadraticModel
+from dimod.binary.binary_quadratic_model import quicksum
+from dimod.quadratic import Integer 
+
+def magic_square(size : int, power : int = 1) -> ConstrainedQuadraticModel: 
+    """Generate a magic square of a particular size. Also can generate a magic square of squares.
+
+    Args:
+        size (int): The sidelength of the magic square, the magic square will have size * size integer variables. 
+        power (int, optional): If set to 1 the problem is a normal magic square, if set to 2 it is a magic square of squares. Must be 1 or 2. Defaults to 1.
+
+    Returns:
+        dimod.ConstrainedQuadraticModel: 
+            A coonstrained quadratic model which represents a magic square. The variables are labeled 
+            by "var_i_j" where i and j are two non-negative integers which represent their column and row respectively.
+            If power is greater than 1 then the value of the variable represents the base, so the magic square is satisfied 
+            by var_i_j**power . 
+    """
+    
+    if power not in [1, 2]: 
+        raise ValueError(f"power must be either 1 or 2, recieved {power}")
+    
+    variables = {}
+    for i, j in product(range(size), range(size)): 
+        variables[(i,j)] = Integer(f"var_{i}_{j}", lower_bound= 1)
+    
+    # the sum across every row, column, diagonal
+    constraint_sum = Integer(f"sum", lower_bound= 1)
+    
+    cqm = ConstrainedQuadraticModel()
+    
+    for i in range(size):
+        # Every row must equal the same value
+        cqm.add_constraint_from_comparison(quicksum([variables[(i,j)]**power for j in range(size)]) - constraint_sum == 0)
```

Comment:
Can also using dict's conversion from tuple indexing
```suggestion
        cqm.add_constraint_from_comparison(quicksum([variables[i, j]**power for j in range(size)]) - constraint_sum == 0)
```
here and elsewhere.